src/app.js

```diff
@@ -1,12 +1,131 @@
+import { AsyncList } from './asyncList.js';
+import { CoffeeShop } from './coffeeShop.js';
+
 /**
- * @param {Object} position
- * @param {Number} position.x
- * @param {Number} position.y
- * 
- * @returns {Array<position>}
+ * @typedef {object} Point
+ * @property {number} x
+ * @property {number} y
  */
-export function getNearestShops(position) {
-  // code
-  
-  return [];
+
+/**
+ * @callback getNearestShops
+ * @param {Point} position
+ * @param {number} numberOfShopsToReturn
+ * @returns {Promise<Array<import('./coffeeShop.js').ShopDistance>>}
+ */
+
+/**
+ * @param {fetchShopList} fetchShops
+ * @param {import("./coffeeShop.js").shopDistanceTo} shopDistanceTo
+ * @param {import("./coffeeShop").distanceAsc} distanceAsc
+ * @returns {getNearestShops}
+ */
+export function nearestShopsFactory(fetchShops, shopDistanceTo, distanceAsc) {
+    return (position, numberOfShopsToReturn) => fetchShops()
+        .map(shopDistanceTo(position))
+        .sort(distanceAsc)
+        .take(numberOfShopsToReturn);
+}
+
+/**
+ * @callback fetchShopList
+ * @returns {import('./asyncList.js').AsyncList}
+ */
+
+/**
+ * @param {fetchAPIToken} fetchAPIToken
+ * @param {fetchCoffeeShops} fetchShops
+ * @returns {fetchShopList}
+ */
+export function fetchShopListFactory(fetchAPIToken, fetchShops) {
+    return () => new AsyncList(fetchAPIToken().then(fetchShops));
 }
+
+/**
+ * @callback fetchAPIToken
+ * @returns {Promise<string>}
+ */
+
+/**
+ * @param {string} host
+ * @param {string} path
+ * @param {import("./io.js").httpFetch<{token?: string}>} httpFetch
+ * @return {fetchAPIToken}
+ */
+export function fetchCoffeeShopAPITokenFactory(host, path, httpFetch) {
+    const ERROR_PREFIX = 'Fetch Token';
+    return () => {
+        const options = {
+            host: host,
+            path: path,
+            method: 'POST'
+        };
+
+        return httpFetch(ERROR_PREFIX, options).then(data => {
+            if (data.token === undefined) {
+                return Promise.reject(new Error(ERROR_PREFIX + ': no token found in response'));
+            }
+            return data.token;
+        });
+    };
+}
```

Comment:
`Why do you want to pass httpFetch here?` I'm new to testing and had problems with tests getting bloated so decided to make it a dependency so I can test it separately. 

`Also, Factory again` Because this function only returns another function I didn't know what to call it. This is the case for all. I thought that because it handles dependencies it feels like the factory pattern.

Reply:
Ah, I see. You call these functions that return functions factories. For the record, this is what the Factory pattern actually looks like: https://refactoring.guru/design-patterns/factory-method.

What you have here are just higher order functions. The word `Factory` made me think you tried to apply the `Factory Pattern`. I don't actually see a need for you to write this function like this, I suppose you wanted to optimize by memoizing the configuration part, but you should have left `options` outside of the returned function if that was the case, and you'd still have to store the returned function somewhere to call it later, so the client code would get bloated a bit.